### PR TITLE
feat: store identities in merk tree

### PR DESCRIPTION
**Pull request by shuplenkov**
## Issue being fixed or feature implemented

need to update Identities repository in order to store data in merk tree

## What was done?

Switched to merk DB to store identities

## How Has This Been Tested?

tests

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

File: lib/identity/IdentityMerkDBRepository.js [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    let db;
+    // const db = transaction ? transaction.db : this.db;
+
+    try {
+      let encodedIdentity;
+
+      if (transaction) {
+        db = transaction.db;
+
+        encodedIdentity = db.get(id);
```
Comment (shumkov) [inline lib/identity/IdentityMerkDBRepository.js:63]: Why wy call methods in transaction DB differently? If we have the same namming you can just swap db instance.

File: lib/merkDb/MerkDbTransaction.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+   * Start new transaction in merk DB
+   *
+   * @return {MerkDbTransaction}
+   */
+  start() {
+    if (this.db) {
+      throw new MerkDBTransactionIsAlreadyStartedError();
+    }
+
+    this.db = new MerkDbWrapper(this.merkDB);
```
Comment (shumkov) [inline lib/merkDb/MerkDbTransaction.js:26]: Let’s call it `MerkDbInMemoryDecorator`?

File: lib/merkDb/MerkDbTransactionWrapper.js [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+  }
+
+  /**
+   *
+   * @param {Buffer} key
+   * @return {null|Buffer}
+   */
+  get(key) {
+    if (this.deleted[key.toString('hex')]) {
+      return null;
```
Comment (shumkov) [inline lib/merkDb/MerkDbTransactionWrapper.js:19]: Throw an error with message ”key not found”

File: lib/merkDb/MerkDbTransactionWrapper.js [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    const value = this.data.get(key.toString('hex'));
```
Comment (shumkov) [inline lib/merkDb/MerkDbTransactionWrapper.js:22]: Can we use Buffers as keys?

File: lib/merkDb/MerkDbTransactionWrapper.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+  }
+
+  /**
+   *
+   * @param {Buffer} key
+   *
+   * @return {MerkDbTransactionWrapper}
+   */
+  delete(key) {
+    this.data.delete(key.toString('hex'));
```
Comment (shumkov) [inline lib/merkDb/MerkDbTransactionWrapper.js:52]: I would try to get (`this.getSync`) it first and then delete it if present just to mimic to merkDb behaviour.

File: lib/merkDb/MerkDbTransactionWrapper.js [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+    return this;
+  }
+
+  /**
+   *
+   * @return {MerkDbTransactionWrapper}
+   */
+  commit() {
+    let batch = this.db.batch();
```
Comment (shumkov) [inline lib/merkDb/MerkDbTransactionWrapper.js:63]: Let’s check first if there something to commit?

**Review by shumkov — CHANGES_REQUESTED** [review]

File: lib/merkDb/MerkDbTransactionWrapper.js (hunk 4 above)
Comment (shuplenkov) [reply lib/merkDb/MerkDbTransactionWrapper.js:22]: No

File: lib/identity/IdentityMerkDBRepository.js (hunk 1 above)
Comment (shuplenkov) [reply lib/identity/IdentityMerkDBRepository.js:63]: differently from what?

File: lib/merkDb/MerkDbTransaction.js (hunk 2 above)
Comment (shuplenkov) [reply lib/merkDb/MerkDbTransaction.js:26]: Why "InMemory"?

File: lib/identity/IdentityMerkDBRepository.js (hunk 1 above)
Comment (shumkov) [reply lib/identity/IdentityMerkDBRepository.js:63]: From MerkDB interface.

File: lib/merkDb/MerkDbTransaction.js (hunk 2 above)
Comment (shumkov) [reply lib/merkDb/MerkDbTransaction.js]: Coz we store data in memory

File: .env.example [hunk 7]
```diff
@@ -6,7 +6,7 @@ ABCI_PORT=26658
 BLOCKCHAIN_STATE_LEVEL_DB_FILE=./db/blockchain-state
 
 # Identity level db file
-IDENTITY_LEVEL_DB_FILE=./db/identity
+IDENTITY_MERK_DB_FILE=./db/identity
```
Comment (shumkov) [inline .env.example:9]: Since we have two databases let's specify the DB here: `./db/identities-merkdb`

**Review by shumkov — APPROVED** [review]
Looks good! 👍

**Pull request merged**